pinot-server/src/main/java/org/apache/pinot/server/starter/helix/OffsetBasedConsumptionStatusChecker.java

```diff
@@ -0,0 +1,154 @@
+        return false;
+      }
+      SegmentDataManager segmentDataManager = tableDataManager.acquireSegment(segName);
+      if (segmentDataManager == null) {
+        LOGGER.info("SegmentDataManager is not yet setup for segment {}. Will check consumption status later", segName);
+        return false;
+      }
+      if (!(segmentDataManager instanceof LLRealtimeSegmentDataManager)) {
+        // There's a small chance that after getting the list of consuming segment names at the beginning of this method
+        // up to this point, a consuming segment gets converted to a committed segment. In that case status check is
+        // returned as false and in the next round the new consuming segment will be used for fetching offsets.
+        LOGGER.info("Segment {} is already committed. Will check consumption status later", segName);
+        tableDataManager.releaseSegment(segmentDataManager);
+        return false;
+      }
+      LLRealtimeSegmentDataManager rtSegmentDataManager = (LLRealtimeSegmentDataManager) segmentDataManager;
+      StreamPartitionMsgOffset latestIngestedOffset = rtSegmentDataManager.getCurrentOffset();
+      StreamPartitionMsgOffset latestStreamOffset = _segmentNameToLatestStreamOffset.containsKey(segName)
+          ? _segmentNameToLatestStreamOffset.get(segName)
+          : rtSegmentDataManager.fetchLatestStreamOffset();
+      tableDataManager.releaseSegment(segmentDataManager);
+      if (latestStreamOffset == null || latestIngestedOffset == null) {
+        LOGGER.info("Null offset found for segment {} - latest stream offset: {}, latest ingested offset: {}. "
+            + "Will check consumption status later", segName, latestStreamOffset, latestIngestedOffset);
+        return false;
+      }
+      if (latestIngestedOffset.compareTo(latestStreamOffset) < 0) {
+        LOGGER.info("Latest ingested offset {} in segment {} is smaller than stream latest available offset {} ",
+            latestIngestedOffset, segName, latestStreamOffset);
+        _segmentNameToLatestStreamOffset.put(segName, latestStreamOffset);
+        allSegsReachedLatest = false;
+        continue;
+      }
+      LOGGER.info("Segment {} with latest ingested offset {} has caught up to the latest stream offset {}", segName,
+            latestIngestedOffset, latestStreamOffset);
+      _alreadyProcessedSegments.add(segName);
+    }
+    return allSegsReachedLatest;
+  }
+
+  private TableDataManager getTableDataManager(String segmentName) {
+    LLCSegmentName llcSegmentName = new LLCSegmentName(segmentName);
+    String tableName = llcSegmentName.getTableName();
+    String tableNameWithType = TableNameBuilder.forType(TableType.REALTIME).tableNameWithType(tableName);
+    return _instanceDataManager.getTableDataManager(tableNameWithType);
+  }
+
+  private static Set<String> findConsumingSegments(HelixAdmin helixAdmin, String helixClusterName, String instanceId) {
```

Comment:
Why don't we query the local `RealtimeTableDataManager` instead of getting all the tables and filtering out only those that have this instance for some of the partitions?

Reply:
I actually wanted to do this originally, but then realized that `TableDataManager`s and `SegmentDataManager`s get added when there's a helix transition message for a segment to go in consuming state from offline state. If for some reason, helix transition messages for some segments are temporarily lost or delayed at startup, then their segmentDataManagers (or even tableDataManagers) won't be available. Ideal state in ZK, on the other hand, is the source of truth.
That being said, I can add one more check before getting ideal state for every table and that's checking if the resource is indeed a realtime table.